[shard] Enable ShardedTensor as a nn.Parameter

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78590

Do whatever it takes to make ShardedTensor be able to be wrapped by
nn.Parameter, so that it appears to the module.parameters() and
module.named_parameters() list. We could in the future remove the
named_params_with_sharded_tensor API.

wanchaol committed:
[shard] Enable ShardedTensor as a nn.Parameter

Do whatever it takes to make ShardedTensor be able to be wrapped by
nn.Parameter, so that it appears to the module.parameters() and
module.named_parameters() list. We could in the future remove the
named_params_with_sharded_tensor API.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -425,6 +425,26 @@ def cpu(
+        If this ShardedTensor is already on CPU memory, then no copy is
+        performed and original object is returned.
+
+        .. note:: When moving a ShardedTensor from GPU to CPU, the ShardedTensor might
```
Curious why this doc is here, as this API should just move CPU -> GPU?

rohan-varma commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -425,6 +425,26 @@ def cpu(
+        """
+        Returns a copy of this object in CUDA memory.
+
+        If this ShardedTensor is already on CPU memory, then no copy is
```
So we'd return the CPU sharded tensor as-is, but this seems contradictory to `.cuda()` call?

wanchaol replied:
sorry forget to update this call to the correct semantic, updating now.

wanchaol committed:
Update on "[shard] Enable ShardedTensor as a nn.Parameter"


Do whatever it takes to make ShardedTensor be able to be wrapped by
nn.Parameter, so that it appears to the module.parameters() and
module.named_parameters() list. We could in the future remove the
named_params_with_sharded_tensor API.

[ghstack-poisoned]

wanchaol committed:
Update on "[shard] Enable ShardedTensor as a nn.Parameter"


Do whatever it takes to make ShardedTensor be able to be wrapped by
nn.Parameter, so that it appears to the module.parameters() and
module.named_parameters() list. We could in the future remove the
named_params_with_sharded_tensor API.

[ghstack-poisoned]

wanchaol committed:
Update on "[shard] Enable ShardedTensor as a nn.Parameter"


Do whatever it takes to make ShardedTensor be able to be wrapped by
nn.Parameter, so that it appears to the module.parameters() and
module.named_parameters() list. We could in the future remove the
named_params_with_sharded_tensor API.

[ghstack-poisoned]

wanchaol commented:
closing this in favor of https://github.com/pytorch/pytorch/pull/79825 instead.
